Add device authorization grant (device code flow - rfc 8628)

duzumaki opened the pull request:
Note to reviewers: I've made this a "commit by commit" pr which means it's easier to review the pr if you go commit by commit rather than look at all files changed at once

<!-- See  https://django-oauth-toolkit.readthedocs.io/en/latest/contributing.html#pull-requests -->
<!-- If there's already an issue that this PR fixes, add that issue number below after 'Fixes #' -->
Fixes https://github.com/jazzband/django-oauth-toolkit/issues/962

follow up from 
https://github.com/oauthlib/oauthlib/pull/881
&
https://github.com/oauthlib/oauthlib/pull/889

## Description of the Change

## Checklist

<!-- Replace '[ ]' with '[x]' to indicate that the checklist item is completed. -->
<!-- You can check the boxes now or later by just clicking on them. -->

- [x] PR only contains one change (considered splitting up PR)
- [x] unit-test added
- [x] documentation updated
- [x] `CHANGELOG.md` updated (only for user relevant changes)
- [x] author name in `AUTHORS`

dopry commented on oauth2_provider/oauth2_backends.py:
```diff
@@ -148,6 +151,16 @@ def create_authorization_response(self, request, scopes, credentials, allow):
         except oauth2.OAuth2Error as error:
             raise OAuthToolkitError(error=error, redirect_uri=credentials["redirect_uri"])
 
+    def create_device_authorization_response(self, request: HttpRequest):
```
Are you sure this is a django.http.HttpRequest and not an oauthlib.common.Request?

dopry reviewed (commented):
This looks excellent, Only one thing grabbed my attention in my cursory code review, the type of the request parameter. Take a moment to double check that type. I've been bitten by OAuthLib's recasting of Request on a number of occasions. I hope to get time to more thoroughly review this by the end of the week

danlamanna reviewed (commented):
This looks awesome! I left some comments even though I'm not a maintainer, I'm just an excited downstream user :). If you're too busy to address any of my feedback let me know, I'd be happy to spend some time on it.

I got this up and running locally and was able to complete the authorization flow. Other than the comments I left inline, I have a few thoughts.

1. Were you planning on adding a default view and template to complete the flow, similar to the way other grant types operate? Obviously the device flow user interaction can be highly customized, but I think a simple view could provide a decent out of the box experience. This was the code I wrote on my application to test this end-to-end:
```python
from oauthlib.oauth2.rfc8628.errors import (
    AccessDenied,
    ExpiredTokenError,
)
from oauth2_provider.models import get_device_model
from django import forms


class DeviceForm(forms.Form):
    user_code = forms.CharField(required=True)


@login_required
def oauth_device_authenticate(request):
    form = DeviceForm(request.POST or None)

    if request.method == "POST" and form.is_valid():
        user_code = form.cleaned_data["user_code"]
        device = get_device_model().objects.filter(user_code=user_code).first()

        if device is None:
            form.add_error("user_code", "Incorrect user code")
        else:
            if timezone.now() > device.expires:
                device.status = device.EXPIRED
                device.save(update_fields=["status"])
                raise ExpiredTokenError

            if device.status in (device.DENIED, device.AUTHORIZED):
                raise AccessDenied

            if device.user_code == user_code:
                device.status = device.AUTHORIZED
                device.save(update_fields=["status"])
                return HttpResponseRedirect(reverse("oauth-device-authenticate-success"))

    return render(request, "device_authenticate.html", {"form": form})


@login_required
def oauth_device_authenticate_success(request):
    return render(request, "device_authenticate_success.html")
```
2. Likewise, are downstreams expected to implement their own `/token` endpoint?

3. Should DOT be a little bit more opinionated about how to generate things like `user_code`? There seems to be a good bit in the RFC (6.1) about best practices that we could encode for downstreams: e.g. using a shorter code with enough entropy that has readable characters and is compared case-insensitively.

Thanks again for all this work :)

duzumaki commented:
> 1. Were you planning on adding a default view and template to complete the flow, similar to the way other grant types operate? Obviously the device flow user interaction can be highly customized, but I think a simple view could provide a decent out of the box experience. This was the code I wrote on my application to test this end-to-end:


> ```python
> from oauthlib.oauth2.rfc8628.errors import (
>     AccessDenied,
>     ExpiredTokenError,
> )
> from oauth2_provider.models import get_device_model
> from django import forms
> 
> 
> class DeviceForm(forms.Form):
>     user_code = forms.CharField(required=True)
> 
> 
> @login_required
> def oauth_device_authenticate(request):
>     form = DeviceForm(request.POST or None)
> 
>     if request.method == "POST" and form.is_valid():
>         user_code = form.cleaned_data["user_code"]
>         device = get_device_model().objects.filter(user_code=user_code).first()
> 
>         if device is None:
>             form.add_error("user_code", "Incorrect user code")
>         else:
>             if timezone.now() > device.expires:
>                 device.status = device.EXPIRED
>                 device.save(update_fields=["status"])
>                 raise ExpiredTokenError
> 
>             if device.status in (device.DENIED, device.AUTHORIZED):
>                 raise AccessDenied
> 
>             if device.user_code == user_code:
>                 device.status = device.AUTHORIZED
>                 device.save(update_fields=["status"])
>                 return HttpResponseRedirect(reverse("oauth-device-authenticate-success"))
> 
>     return render(request, "device_authenticate.html", {"form": form})
> 
> 
> @login_required
> def oauth_device_authenticate_success(request):
>     return render(request, "device_authenticate_success.html")


This code I put in tutotorial_06.rst was a simplified version of how I implemented in my own authserver. 
However, with the device view being highly customizable and specific to your own auth server and some implementations even working with open id connect(which is not part of the rfc), I opted to not include it. For example, I have plans on adding extensions similar to mutual TLS to my device flow in my auth server. 

However this is up to the maintainers to decide but I'd rather get this merged and we add it later if we deem it important as I also worked on making sure oauthlib can support this grant so I've been working on this for quite some time now to put everything in place([this](https://github.com/oauthlib/oauthlib/pull/881) pr & [this](https://github.com/oauthlib/oauthlib/pull/889))


> ```
> 
> 2. Likewise, are downstreams expected to implement their own `/token` endpoint?
No, they can if they want but their auth server should have a working `/token` endpoint as a prerequisite 

> 3. Should DOT be a little bit more opinionated about how to generate things like `user_code`? There seems to be a good bit in the RFC (6.1) about best practices that we could encode for downstreams: e.g. using a shorter code with enough entropy that has readable characters and is compared case-insensitively.
That's why I [updated](https://github.com/oauthlib/oauthlib/pull/885) oauthlib to support the ability to pass in custom user code generator callables if you set the setting I made for it in oauth toolkit. I'm being core RFC focused here first and if anything opinionated needs to be added I think we can add it later, This pr is already chunky as is the way I see it


> Thanks again for all this work :)
Thank you!

duzumaki commented:
@danlamanna 
> 1. Were you planning on adding a default view and template to complete the flow, similar to the way other grant types operate? Obviously the device flow user interaction can be highly customized, but I think a simple view could provide a decent out of the box experience. This was the code I wrote on my application to test this end-to-end:


> ```python
> from oauthlib.oauth2.rfc8628.errors import (
>     AccessDenied,
>     ExpiredTokenError,
> )
> from oauth2_provider.models import get_device_model
> from django import forms
> 
> 
> class DeviceForm(forms.Form):
>     user_code = forms.CharField(required=True)
> 
> 
> @login_required
> def oauth_device_authenticate(request):
>     form = DeviceForm(request.POST or None)
> 
>     if request.method == "POST" and form.is_valid():
>         user_code = form.cleaned_data["user_code"]
>         device = get_device_model().objects.filter(user_code=user_code).first()
> 
>         if device is None:
>             form.add_error("user_code", "Incorrect user code")
>         else:
>             if timezone.now() > device.expires:
>                 device.status = device.EXPIRED
>                 device.save(update_fields=["status"])
>                 raise ExpiredTokenError
> 
>             if device.status in (device.DENIED, device.AUTHORIZED):
>                 raise AccessDenied
> 
>             if device.user_code == user_code:
>                 device.status = device.AUTHORIZED
>                 device.save(update_fields=["status"])
>                 return HttpResponseRedirect(reverse("oauth-device-authenticate-success"))
> 
>     return render(request, "device_authenticate.html", {"form": form})
> 
> 
> @login_required
> def oauth_device_authenticate_success(request):
>     return render(request, "device_authenticate_success.html")


This code I put in tutotorial_06.rst was a simplified version of how I implemented in my own authserver. 
However, with the device view being highly customizable and specific to your own auth server and some implementations even working with open id connect(which is not part of the rfc), I opted to not include it. For example, I have plans on adding extensions similar to mutual TLS to my device flow in my auth server. 

However this is up to the maintainers to decide but I'd rather get this merged and we add it later if we deem it important as I also worked on making sure oauthlib can support this grant so I've been working on this for quite some time now to put everything in place([this](https://github.com/oauthlib/oauthlib/pull/881) pr & [this](https://github.com/oauthlib/oauthlib/pull/889))


> ```
> 
> 2. Likewise, are downstreams expected to implement their own `/token` endpoint?
No, they can if they want but their auth server should have a working `/token` endpoint as a prerequisite 

> 3. Should DOT be a little bit more opinionated about how to generate things like `user_code`? There seems to be a good bit in the RFC (6.1) about best practices that we could encode for downstreams: e.g. using a shorter code with enough entropy that has readable characters and is compared case-insensitively.
That's why I [updated](https://github.com/oauthlib/oauthlib/pull/885) oauthlib to support the ability to pass in custom user code generator callables if you set the setting I made for it in oauth toolkit. I'm being core RFC focused here first and if anything opinionated needs to be added I think we can add it later, This pr is already chunky as is the way I see it


> Thanks again for all this work :)
Thank you!

duzumaki commented on oauth2_provider/models.py:
```diff
@@ -650,11 +654,93 @@ class Meta(AbstractIDToken.Meta):
+    status = models.CharField(
+        max_length=64, blank=True, choices=DEVICE_FLOW_STATUS, default=AUTHORIZATION_PENDING
+    )
+    client_id = models.CharField(max_length=100, default=generate_client_id, db_index=True)
```
Yeah this default shouldn't be needed

duzumaki commented on oauth2_provider/models.py:
```diff
@@ -650,11 +654,93 @@ class Meta(AbstractIDToken.Meta):
+    id = models.BigAutoField(primary_key=True)
+    device_code = models.CharField(max_length=100, unique=True)
+    user_code = models.CharField(max_length=100)
+    scope = models.CharField(max_length=64, default="openid")
```
This might be a remnant from when I first started this work; I wanted to integrate it with openid but decided to keep it just focused to oauth2, the official rfc. I'll double check this and try remove it. Can't recall if there was another reason I left it here

duzumaki replied:
This might be a remnant from when I first started this work; I wanted to integrate it with openid but decided to keep it just focused to oauth2, the official rfc. I'll double check this and try remove it. also [this](https://github.com/jazzband/django-oauth-toolkit/pull/1539/files/69b702099bf5ed028d9ff5b3490f1316ad55d163#r1910117644)

duzumaki commented on oauth2_provider/models.py:
```diff
@@ -650,11 +654,93 @@ class Meta(AbstractIDToken.Meta):
+
+@dataclass
+class DeviceCodeResponse:
+    verification_uri: str
```
Yeah can do

dopry replied:
Is this a PR pending upstream? Should we wait for that to land before merging this?

duzumaki replied:
Oauthlib is still due a release even so this was never going to be merged first

duzumaki replied:
verification_uri_complete will now be [returned](https://github.com/jazzband/django-oauth-toolkit/pull/1539/files#diff-1ebdc91fdacce3c6c243c059421477ecd84b80bbc0381a1d21b424a66135241fR737) if [this](https://github.com/jazzband/django-oauth-toolkit/pull/1539/files#diff-bcd09bd2f89c4ab670e9facf303e0a81a6348422a4bb07ff1eb0c2b7b4466e93R46) is set

I've set it for the test idp [here](https://github.com/jazzband/django-oauth-toolkit/pull/1539/files#diff-a886518d2d9dcc1bbd9bf4f98fcf6f132c8f31ec518de147ebdeace4df04ef63R207) if one wants to see it locally

duzumaki commented on oauth2_provider/models.py:
```diff
@@ -650,11 +654,93 @@ class Meta(AbstractIDToken.Meta):
         swappable = "OAUTH2_PROVIDER_ID_TOKEN_MODEL"
 
 
+class AbstractDevice(models.Model):
```
[It's not the grant](https://github.com/jazzband/django-oauth-toolkit/pull/1539/files/69b702099bf5ed028d9ff5b3490f1316ad55d163#r1910094117)

duzumaki commented on oauth2_provider/views/device.py:
```diff
@@ -0,0 +1,33 @@
+        headers, response, status = self.create_device_authorization_response(request)
+
+        device_request = DeviceRequest(
+            client_id=request.POST["client_id"], scope=request.POST.get("scope", oauth2_settings.READ_SCOPE)
```
I'll have to double check this because I remember when I first start this work something in oauthlib failing since no scope was present and it wasn't a check I added there

duzumaki commented on oauth2_provider/views/mixins.py:
```diff
@@ -114,6 +114,20 @@ def create_authorization_response(self, request, scopes, credentials, allow):
+        instance.
+        :param request: The current django.http.HttpRequest object
+        """
+        oauth2_settings.EXTRA_SERVER_KWARGS = {
```
Looks like a better spot yup. I'll test it out. Didn't see that method until now

duzumaki commented on tests/test_device.py:
```diff
@@ -0,0 +1,155 @@
+            name="test_client_credentials_app",
+            user=cls.dev_user,
+            client_type=Application.CLIENT_PUBLIC,
+            authorization_grant_type=Application.GRANT_CLIENT_CREDENTIALS,
```
It can be put it doesn't matter for this test. this is the initial device autothirzation stage.
Grant type is only checked when the device attempts to make a call to the authserver and/or the user tries to go to the approve deny page

dopry replied:
please correct, other test cases may inherit and it should be set properly.

duzumaki commented on docs/tutorial/tutorial_06.rst:
```diff
@@ -0,0 +1,144 @@
+        # there should only be one
+        device: Device = get_device_model().objects.get(user_code=user_code)
+
+        if datetime.now(tz=UTC) > device.expires:
```
I had this originally and don't remember why I removed it (been on and off this branch for the last 2-3 months so forgot some of my initial decisions). I'll play around with adding it back in and maybe re-jog why I removed it (or not)

duzumaki replied:
added [here](https://github.com/jazzband/django-oauth-toolkit/pull/1539/files#diff-1ebdc91fdacce3c6c243c059421477ecd84b80bbc0381a1d21b424a66135241fR691)

and used [here](https://github.com/jazzband/django-oauth-toolkit/pull/1539/files#diff-d3838f55b2634a980bb2b72fe69c9f8007587550679e0da44522bbb18533f4e1R66)

dopry commented on tests/test_device.py:
```diff
@@ -0,0 +1,155 @@
+
+@pytest.mark.usefixtures("oauth2_settings")
+@pytest.mark.oauth2_settings(presets.DEFAULT_SCOPES_RW)
+class BaseTest(TestCase):
```
please rename DeviceFlowBaseTestCase

duzumaki replied:
https://github.com/jazzband/django-oauth-toolkit/pull/1539/files#diff-72107d6f4a703abaeb6a7cf709e0c99f7ed0b7c74f0b14b0cbc4f35c7c650e26R25

dopry commented:
@danlamanna thanks for putting it through it's paces and for the code review. We always appreciate extra hands in the community kicking the tires on pull requests.  

@duzumaki I haven't had a chance to get into a thorough review yet. It's high on my OSS priority list. It would be nice to have a working implementation in the example idp/rp in tests/app. If you need any help on the rp side there, I'm happy to lend a hand. That will reduce our testing overhead as maintainers. It's a lot to review an OAuth Flow without also having to implement part of it as well, especially as we haven't been as awash in the specification as you seem to have been for a bit.  I am partial to the idea of having necessary default views in DOT, I really prefer as much of  a batteries included experience for our users.  If we give people a half implementation in an initial release it will be a lot of work for a lot of people, then when we add in our own view implementations it'll be an upgrade headache for all of those users. If we can deliver a view that adheres to best practice with reasonable defaults which users can override I would much prefer that.

duzumaki commented:
@dopry 

makes sense. i'll move some views i had in my own custom auth server and port them over to this pr

duzumaki commented:
@dopry 

makes sense. i'll move some implementation i had in my own custom auth server and port them over to this pr

duzumaki commented:
@dopry am I missing something?

I start the svelte thing and just see this. buttons don't work. I've added the views to the ldp path locally but can't get the token endpoint to work because the refresh token model wants to be saved with the django user (from request.user) but it's never populated in any view. Is there some setting missing or something in idp/settings?

<img width="328" alt="image" src="https://github.com/user-attachments/assets/c27f82e3-402a-4ad0-983d-ed70cc0e2e42" />

duzumaki commented:
@dopry Okay I have a working implementation locally in the idp folder and added the necessary views, templates and form. I'll clean it up and push it tomorrow. I have to write some tests before I push it and i'm done for the work day.

There's also a change to the abstract refresh token that I had to make as well. I'll explain it tomorrow in the commits 👌

n2ygk commented:
@duzumaki It looks like you may be battling with pre-commit which is fixing your code formatting after push. Do you have pre-commit installed locally? See https://django-oauth-toolkit.readthedocs.io/en/latest/contributing.html#code-style

duzumaki commented:
@n2ygk The pushes aren't because of the pre-commit. I use rebase so I'm fixing the history so it's easier to review

dopry commented:
I'll take a look later this evening

duzumaki commented:
hey @dopry managed to give it a look? understand you may be busy

dopry commented:
We're moving in the right direction. I'd love to see those open comments get resolved.

dopry commented on docs/tutorial/tutorial_06.rst:
```diff
@@ -0,0 +1,95 @@
+
+.. code-block:: sh
+
+    cd tests/app/idp
```
Why are we navigating to this directory?

duzumaki replied:
Doc is missing a `python manage.py runserver` after this line

duzumaki commented:
@dopry @danlamanna 

> We're moving in the right direction. I'd love to see those open comments get resolved.

just addressed all open comments

danlamanna reviewed (commented):
Thanks for taking another pass at this. After taking a closer look, I noticed that the tokens being issued have `null` user ids, so I don't think things are working quite correctly out of the box.

I saw your message on commit [`1167cd6` (#1539)](https://github.com/jazzband/django-oauth-toolkit/pull/1539/commits/1167cd6e3a356fdfe371539164e760bd6bbd9748) and I think that this is the crux of the problem. For other flows, like auth code, the user can also be absent. The way it gets set on `request.user` is via DOT [extracting it from the Grant](https://github.com/duzumaki/django-oauth-toolkit/blob/a9eb10e390af030e5813cd285e0bf735e90c80b5/oauth2_provider/oauth2_validators.py#L493) and overriding it. What I'm proposing is that this flow does something similar and creates the token with the user retrieved from the Grant (now named `Device`). The one caveat is that the `DeviceGrant` would need to have a nullable `User` unlike other grants, since the user isn't known when it's initially created.

dopry commented on oauth2_provider/views/base.py:
```diff
@@ -307,6 +318,38 @@ def post(self, request, *args, **kwargs):
             response[k] = v
         return response
 
+    @method_decorator(ratelimit(key="ip", rate=f"1/{oauth2_settings.DEVICE_FLOW_INTERVAL}"))
```
I would prefer not to add any unnecessary dependencies. Every dep we add is something we need to track and support to some extent. If rate limiting isn't a requirement and isn't listed in a best practice specification somewhere I would lean towards removing it here.

duzumaki replied:
```
Due to the polling nature of this protocol (as specified in
   [Section 3.4](https://datatracker.ietf.org/doc/html/rfc8628#section-3.4)), care is needed to avoid overloading the capacity of the
   token endpoint.  To avoid unneeded requests on the token endpoint,
   the client SHOULD only commence a device authorization request when
   prompted by the user and not automatically, such as when the app
   starts or when the previous authorization session expires or fails.
```
   
 ```
   interval
      OPTIONAL.  The minimum amount of time in seconds that the client
      SHOULD wait between polling requests to the token endpoint.  If no
      value is provided, clients MUST use 5 as the default.
```
   
The way i'm read it is, it's the server's responsibility to enforce the default for the device hence the rate limiting. In my auth server we have it but will remove it here in DOT

duzumaki replied:
@danlamanna ^

duzumaki replied:
[removed](https://github.com/jazzband/django-oauth-toolkit/pull/1539/files#diff-d53bb3f2e759bd99d1793dc47a8356795a13294c788b830d2e01deec86b7498eR320)

duzumaki replied:
@dopry @danlamanna [removed](https://github.com/jazzband/django-oauth-toolkit/pull/1539/files#diff-d53bb3f2e759bd99d1793dc47a8356795a13294c788b830d2e01deec86b7498eR320)

dopry replied:
I read it the opposite. I don't see anything saying the server should or the server must... It's all 'the client should' or client 'must'. I feel the goal of the spec here is for good actors to be considerate.

duzumaki commented on oauth2_provider/views/device.py:
```diff
@@ -0,0 +1,103 @@
+        return render(request, "oauth2_provider/device/user_code.html", {"form": form})
+
+    user_code: str = form.cleaned_data["user_code"]
+    device: Device = get_device_model().objects.get(user_code=user_code)
```
Using filter here is semantically different than using .get()
I'm saying "there should only be 1  device using this code duirng any active device session in the auth server" 

It's better the exception is caught here

dopry replied:
Is there a unique constraint on the schema?  I feel like this is too late to get the error.  What should the response be if no code is found in this scenario? Which part of the spec covers it?

duzumaki replied:
There's a unique constraint on the device code (following the RFC) but nothing mentioned uniqueness of the user code.

However in practice this is a problem(that the rfc should mention but doesn't)

The original version of this pr had the constraint but i removed it since it isn't part of the rfc.
I can add the constraint back in but it's up to the downstream user to inform the device to send another request
and do any other custom logic like making the user code only unique for the duration device flow session

dopry replied:
I agree. From a UX perspective it is important that we return an appropriate error so clients can properly inform users about how to proceed. This is a form handler, so in theory wouldn't we catch this in form validation? Maybe we should render the form with an error here.

duzumaki commented:
@danlamanna 

> What I'm proposing is that this flow does something similar and creates the token with the user retrieved from the Grant (now named Device). The one caveat is that the DeviceGrant would need to have a nullable User unlike other grants, since the user isn't known when it's initially created.

Yeah this is what I said I do in my own auth server in the [commit message](https://github.com/jazzband/django-oauth-toolkit/pull/1539/commits/1167cd6e3a356fdfe371539164e760bd6bbd9748)


```

DOT currently assume the user will be derived from the django request.user object (from the logic throughout DOT, not the model itself).

Since the device flow happens out of band there is no request.user available when the call to token is made, we have to make this field none.

How do I handle it in my own custom auth server:
In my custom auth server how I associate a refresh token with a user is to have a field (column in the refresh token table) that has the payload of the original JWT what was made when the refresh token was issued and I use the sub claim in the payload to know “this user has the refresh token” which prevents it relying on django solely for the user information but the stateless JWT instead
```

duzumaki commented on oauth2_provider/views/device.py:
```diff
@@ -0,0 +1,103 @@
+        return render(request, "oauth2_provider/device/user_code.html", {"form": form})
+
+    user_code: str = form.cleaned_data["user_code"]
+    device: Device = get_device_model().objects.get(user_code=user_code)
```
I'm not following. when `OAUTH_DEVICE_USER_CODE_GENERATOR` is set that's the user code that will be used, it's up to downstream to determine how they want their user code to be

duzumaki commented on tests/test_device.py:
```diff
@@ -0,0 +1,269 @@
+        assert post_response_valid["Location"] == device_confirm_url
+
+        device_confirm_url = reverse("oauth2_provider:device-confirm", kwargs={"device_code": "abc"})
+        assert post_response_valid["Location"] == device_confirm_url
```
[removed](https://github.com/jazzband/django-oauth-toolkit/pull/1539/files#diff-72107d6f4a703abaeb6a7cf709e0c99f7ed0b7c74f0b14b0cbc4f35c7c650e26R170-R173)

duzumaki commented on oauth2_provider/templates/oauth2_provider/device/user_code.html:
```diff
@@ -0,0 +1,15 @@
+<!-- device_verification.html -->
```
[removed](https://github.com/jazzband/django-oauth-toolkit/pull/1539/files#diff-874a68dc302c9da1a731379ad57b51216574ce2f9b81cc3fbee44f1cb194ad89R1)

duzumaki commented on oauth2_provider/settings.py:
```diff
@@ -39,6 +42,8 @@
     "CLIENT_SECRET_HASHER": "default",
     "ACCESS_TOKEN_GENERATOR": None,
+    "OAUTH_DEVICE_VERIFICATION_URI": None,
+    "OAUTH_DEVICE_USER_CODE_GENERATOR": lambda: user_code_generator(),
```
https://github.com/jazzband/django-oauth-toolkit/pull/1539/files#diff-bcd09bd2f89c4ab670e9facf303e0a81a6348422a4bb07ff1eb0c2b7b4466e93R46

duzumaki commented on tests/app/idp/idp/settings.py:
```diff
@@ -199,6 +201,8 @@
 OAUTH2_PROVIDER = {
     "OAUTH2_VALIDATOR_CLASS": "idp.oauth.CustomOAuth2Validator",
+    "OAUTH_DEVICE_VERIFICATION_URI": "http://127.0.0.1:8000/o/device",
+    "OAUTH_DEVICE_USER_CODE_GENERATOR": lambda: user_code_generator(),
```
https://github.com/jazzband/django-oauth-toolkit/pull/1539/files#diff-a886518d2d9dcc1bbd9bf4f98fcf6f132c8f31ec518de147ebdeace4df04ef63R205

duzumaki commented on oauth2_provider/templates/oauth2_provider/device/accept_deny.html:
```diff
@@ -0,0 +1,14 @@
+<!DOCTYPE html>
```
[accept deny using oauth2 provder html base](https://github.com/jazzband/django-oauth-toolkit/pull/1539/files#diff-496db2f54c74503da1bf20fd5806513e9173c341ce53f76c84c77ad5318f76d4R1)
[user code form using base](https://github.com/jazzband/django-oauth-toolkit/pull/1539/files#diff-874a68dc302c9da1a731379ad57b51216574ce2f9b81cc3fbee44f1cb194ad89R1)

<img width="748" alt="image" src="https://github.com/user-attachments/assets/402ba08f-68ad-4929-bef2-159532359ddd" />

duzumaki commented:
> I saw your message on commit [1167cd6 (#1539)](https://github.com/jazzband/django-oauth-toolkit/pull/1539/commits/1167cd6e3a356fdfe371539164e760bd6bbd9748) and I think that this is the crux of the problem. For other flows, like auth code, the user can also be absent. The way it gets set on request.user is via DOT [extracting it from the Grant](https://github.com/duzumaki/django-oauth-toolkit/blob/a9eb10e390af030e5813cd285e0bf735e90c80b5/oauth2_provider/oauth2_validators.py#L493) and overriding it. What I'm proposing is that this flow does something similar and creates the token with the user retrieved from the Grant (now named Device). The one caveat is that the DeviceGrant would need to have a nullable User unlike other grants, since the user isn't known when it's initially created.

I agree. I'll have this change added tomorrow.
Since we're required to login when the user visits /device in the browser the device user column(to be added) can be set there and then probably [here](https://github.com/duzumaki/django-oauth-toolkit/blob/918f1e849653ee051fe5c614b37f71e4eca4cdfb/oauth2_provider/oauth2_validators.py#L507) we can set 
  request.user = device.user (where request is of type Oauthlib


Just a side note; since oauthlib and DOT have no typing it makes understanding the flow of what's calling what quite cumbersome. I saw that line you pointed where `request.user = grant.user` before and only realised now it's not a `WSGI` object but an ouathlib `Request`. The lack of typing is throwing me off

duzumaki commented on oauth2_provider/models.py:
```diff
@@ -650,11 +654,93 @@ class Meta(AbstractIDToken.Meta):
+        return self.get(client_id=client_id, device_code=device_code, user_code=user_code)
+
+
+class Device(AbstractDevice):
```
> But in django-oauth-toolkit this represents the pending flow that gets persisted.

This seems to be what I'm saying as well. Because it represents the device session and uses the device grant in that session.

Am I misunderstanding grant in Oauth2? I define it as a single object with that is of some type but doesn't have anything to do with state/session.

The `Grant` in `DOT` is for the authorization code flow. It may have similar columns for its table but that's not the full story with the device here

duzumaki commented:
this pr will support adding `verification_uri_complete` to the response and 
allow me to associate the device user in django with the user in the oauthlib request

I'll then update DOT with a user column for the device

https://github.com/oauthlib/oauthlib/pull/891

duzumaki commented:
@danlamanna @dopry 

the access and refresh tokens now set their user foreing keys based off the user of the device flow
added [this](https://github.com/jazzband/django-oauth-toolkit/pull/1539/commits/6e818165eda890cb65f3d79ca0865864fb487f1f) commit

- device model updated [here](https://github.com/jazzband/django-oauth-toolkit/pull/1539/commits/a91dffa7c8e6b29b7534fcabf44d014fed096294#diff-1ebdc91fdacce3c6c243c059421477ecd84b80bbc0381a1d21b424a66135241fR681) with user fk
-  [migration](https://github.com/jazzband/django-oauth-toolkit/pull/1539/commits/2185f8e5ca1722a545804fb1c15f8c208dbdd2d8#diff-088fe693df59fbf5e6f15b10cbbb31dd38e3b146a93bdfa1e7070cb47fe1d7dbR33)
- [where the django request.user is used to set that fk](https://github.com/jazzband/django-oauth-toolkit/pull/1539/commits/6e818165eda890cb65f3d79ca0865864fb487f1f#diff-d3838f55b2634a980bb2b72fe69c9f8007587550679e0da44522bbb18533f4e1R62-R63)
- the [function](https://github.com/jazzband/django-oauth-toolkit/pull/1539/commits/6e818165eda890cb65f3d79ca0865864fb487f1f#diff-81e62c4239be18d2d3320260122965d6a179a9518f08ec95fab24e5233f4a955R81) responsible for setting the oauthlib request object to the device user
- [how DOT tells oauthlib to run that function](https://github.com/jazzband/django-oauth-toolkit/pull/1539/commits/6e818165eda890cb65f3d79ca0865864fb487f1f#diff-bcd09bd2f89c4ab670e9facf303e0a81a6348422a4bb07ff1eb0c2b7b4466e93R284)
- updated [test](https://github.com/jazzband/django-oauth-toolkit/pull/1539/commits/6e818165eda890cb65f3d79ca0865864fb487f1f#diff-d3838f55b2634a980bb2b72fe69c9f8007587550679e0da44522bbb18533f4e1R62-R63) confirming access and refresh token have the device user

duzumaki commented on oauth2_provider/settings.py:
```diff
@@ -276,8 +278,10 @@ def server_kwargs(self):
+                ("verification_uri_complete", "OAUTH_DEVICE_VERIFICATION_URI_COMPLETE"),
                 ("interval", "DEVICE_FLOW_INTERVAL"),
                 ("user_code_generator", "OAUTH_DEVICE_USER_CODE_GENERATOR"),
+                ("pre_token", "OAUTH_PRE_TOKEN_VALIDATION"),
```
pre_token is read by oauthlib
and the alidator executes [here](https://github.com/oauthlib/oauthlib/blob/028a8f18808bebf3659388f914fc6eb89f9d2bda/oauthlib/oauth2/rfc8628/grant_types/device_code.py#L41)

danlamanna reviewed (commented):
This is looking good to me, I think there's only a couple of outstanding questions/concerns:

1. I noticed the scopes aren't properly being passed on to the token. If I initiate the authorization flow with scope `identity`, it properly shows up for the `Device` but the tokens issued have the `read write` scope.
2. The flow is still raising uncaught exceptions triggering 500 errors when it should be 400 e.g. `raise AuthorizationPendingError` and `raise AccessDenied`.
3. I think the `OAUTH_DEVICE_VERIFICATION_URI_*` settings might be a mistake. Given a single codebase gets deployed at multiple locations, would it be better to derive the URL via the `request` object and `request.build_absolute_uri`?

duzumaki commented:
> I think the OAUTH_DEVICE_VERIFICATION_URI_* settings might be a mistake. Given a single codebase gets deployed at multiple locations, would it be better to derive the URL via the request object and request.build_absolute_uri?

don't see why an env var isn't sufficient here, if the code base is deployed in multiple locations than it's just a matter of those servers having different env vars sets

duzumaki commented:
> I noticed the scopes aren't properly being passed on to the token. If I initiate the authorization flow with scope identity, it properly shows up for the Device but the tokens issued have the read write scope.

It is. you aren't setting the SCOPES setting correctly, I don't think identity is a default anywhere in DOT in that setting

duzumaki commented:
> The flow is still raising uncaught exceptions triggering 500 errors when it should be 400 e.g. raise AuthorizationPendingError and raise AccessDenied.

[Fixed](https://github.com/jazzband/django-oauth-toolkit/pull/1539/files#diff-d53bb3f2e759bd99d1793dc47a8356795a13294c788b830d2e01deec86b7498eR325-R337)
<img width="660" alt="image" src="https://github.com/user-attachments/assets/203def9e-c111-4662-ab1b-e177f6d60162" />

duzumaki commented:
@danlamanna 

> The flow is still raising uncaught exceptions triggering 500 errors when it should be 400 e.g. raise AuthorizationPendingError and raise AccessDenied.

[Fixed](https://github.com/jazzband/django-oauth-toolkit/pull/1539/files#diff-d53bb3f2e759bd99d1793dc47a8356795a13294c788b830d2e01deec86b7498eR325-R337)
<img width="660" alt="image" src="https://github.com/user-attachments/assets/203def9e-c111-4662-ab1b-e177f6d60162" />

duzumaki commented on oauth2_provider/views/base.py:
```diff
@@ -307,6 +318,38 @@ def post(self, request, *args, **kwargs):
+        device = Device.objects.get(device_code=device_code)
+
+        if device.status == device.AUTHORIZATION_PENDING:
+            raise AuthorizationPendingError
```
[Fixed](https://github.com/jazzband/django-oauth-toolkit/pull/1539/files#diff-d53bb3f2e759bd99d1793dc47a8356795a13294c788b830d2e01deec86b7498eR325-R337)
<img width="660" alt="image" src="https://github.com/user-attachments/assets/203def9e-c111-4662-ab1b-e177f6d60162" />

duzumaki commented:
@danlamanna  @dopry 
> I noticed the scopes aren't properly being passed on to the token. If I initiate the authorization flow with scope identity, it properly shows up for the Device but the tokens issued have the read write scope.

It is. you aren't setting the SCOPES setting correctly, I don't think `identity` is even in the repo

<img width="1261" alt="image" src="https://github.com/user-attachments/assets/ff0d5ab6-1bb5-4bf4-b7a3-a5a6eb2a1641" />

---------

> I think the OAUTH_DEVICE_VERIFICATION_URI_* settings might be a mistake. Given a single codebase gets deployed at multiple locations, would it be better to derive the URL via the request object and request.build_absolute_uri?

don't see why an env var isn't sufficient here, if the code base is deployed in multiple locations then it's just a matter of those servers having different env vars set which from the perspective of someone implementing their own auth server, is easier to do

plus something has to tell oauthlib what string to use in the response for the `verification_uri` regardless, I don't see how we can derive it from the request object if someone implementing their own custom auth server can make their own custom device verification endpoint with custom naming

also passing this information down via an env var is familiar practice since we already send other data down to oauthlib via env vars already

-----------

> The flow is still raising uncaught exceptions triggering 500 errors when it should be 400 e.g. raise AuthorizationPendingError and raise AccessDenied.

[Fixed](https://github.com/jazzband/django-oauth-toolkit/pull/1539/files#diff-d53bb3f2e759bd99d1793dc47a8356795a13294c788b830d2e01deec86b7498eR325-R337)
<img width="660" alt="image" src="https://github.com/user-attachments/assets/203def9e-c111-4662-ab1b-e177f6d60162" />

danlamanna reviewed (commented):
You're right, I had my scopes misconfigured :confused:. I added a comment about refresh tokens but otherwise this is looking good to me :)

duzumaki commented on oauth2_provider/oauth2_validators.py:
```diff
@@ -56,6 +56,7 @@
         AbstractApplication.GRANT_CLIENT_CREDENTIALS,
         AbstractApplication.GRANT_OPENID_HYBRID,
     ),
+    "urn:ietf:params:oauth:grant-type:device_code": (AbstractApplication.GRANT_DEVICE_CODE,),
```
Could catch, I had it locally which is how my refresh token calls are working from the device response
pushing it now

<img width="621" alt="image" src="https://github.com/user-attachments/assets/b120a052-f0dc-4927-a381-a4d834f05761" />

duzumaki replied:
[Changed here](https://github.com/jazzband/django-oauth-toolkit/pull/1539/commits/5039617a693ea7f961fdd542209c0c1575431a36)
@danlamanna

duzumaki commented:
@danlamanna  if that refresh token is the last change do we agree this pr is ready for @dopry @n2ygk to give it a look so we can get it merged some time this week?

danlamanna commented:
> @danlamanna if that refresh token commnet is the last change do we agree this pr is ready for @dopry @n2ygk to give it a look so we can get it merged some time next week? (oauthlib will of course need a release so I can get that pr ready for oauthlib)

Yes, LGTM. Thanks for slogging through this with me :)

dopry commented:
@duzumaki  thank you for all your work an attention to this. @danlamanna  Thank you for helping with the review and testing load. We really appreciate you working together to tighten the implementation up. I'm going to give it a thorough run down this weekend and cross reference against the spec, but I suspect we're really close.

duzumaki commented:
hey @dopry any update 🙏

dopry commented:
Hey sorry. I got busy with work and life. Thanks for poking me. I'll try to set aside some time this evening.

dopry commented:
I want to apologize for not getting to this yet. Work has been hectic.  It's still on my radar, I'm just struggling to get time for a thorough review.

cristiprg commented on docs/tutorial/tutorial_06.rst:
```diff
@@ -0,0 +1,100 @@
+Device authorization grant flow
```
```suggestion
Part 6 - Device authorization grant flow
```

small thing, but I checked the rendered docs, and the title needs "Part X -", to be consistent with the other tutorial pages.

<img width="463" alt="image" src="https://github.com/user-attachments/assets/d9d82d34-2f71-4b40-b375-c7f3487cc17c" />

cristiprg commented on docs/tutorial/tutorial_06.rst:
```diff
@@ -0,0 +1,100 @@
+
+    curl --location 'http://127.0.0.1:8000/o/device-authorization/' \
+        --header 'Content-Type: application/x-www-form-urlencoded' \
+        --data-urlencode 'client_id={your application\'s client id}'
```
```suggestion
        --data-urlencode 'client_id={your application\'s client id}
```
there's an extra `'` character. Small thing, but can be annoying when copy/pasting commands from the tutorial 😁

cristiprg commented on docs/tutorial/tutorial_06.rst:
```diff
@@ -0,0 +1,100 @@
+-----------------------
+Device polling occurs concurrently while the user approves or denies the request.
+
+.. image:: _images/device-approve-deny.png
```
♻️ same comment as above regarding the path

<img width="636" alt="image" src="https://github.com/user-attachments/assets/a23f3fd4-39ab-42f5-bd21-6be29676e10d" />

cristiprg commented on docs/tutorial/tutorial_06.rst:
```diff
@@ -0,0 +1,100 @@
+
+Device polling
+--------------
+Note: You should already have the `/token` endpoint implemented in your authorization server before this.
```
Since `/token` is already part of the idp test app, not sure if this statement is really necessary, it only makes the readers following the tutorial wonder whether they've missed something.

cristiprg commented on oauth2_provider/views/device.py:
```diff
@@ -0,0 +1,102 @@
+    user_code: str = form.cleaned_data["user_code"]
+    device: Device = get_device_model().objects.get(user_code=user_code)
+
+    device.user = request.user
+    device.save(update_fields=["user"])
+
+    if device is None:
```
does this `if` need to go before accessing attributes on `device`? In the situation where `device` is None, `device.user` will cause an `AttributeError`

cristiprg commented on oauth2_provider/views/device.py:
```diff
@@ -0,0 +1,102 @@
+def device_confirm_view(request: http.HttpRequest, device_code: str):
+    device: Device = get_device_model().objects.get(device_code=device_code)
+
+    if device.status in (device.AUTHORIZED, device.DENIED):
```
is the only acceptable `status` here `AUTHORIZATION_PENDING`? Maybe we could make this `if` statement

`if device.status != device.AUTHORIZATION_PENDING:`.

Otherwise I guess one attacker having the `device_code` could confirm even if `device.EXPIRED` 🤔

duzumaki replied:
@cristiprg Doesn't it already check if it's expired before this point + you have to be logged in to get here

@dopry You're right, it shouldn't show in the URI. It's used by the device to poll but shouldn't show

cristiprg replied:
I tried finding how others have done it and found this https://youtu.be/-2P68Z47uKE?feature=shared&t=167. They seem to be a security/identity management company so I assume they know what they're doing 😂  hence I'm gonna update this view to use `client_id`, which makes sense because that is public anyways.

duzumaki replied:
Following up from comment, I think that's what I tried to do originally until I remembered the auth is stateless so you'd also want to pass in the grant id too most likely to this view

cristiprg commented on oauth2_provider/views/device.py:
```diff
@@ -0,0 +1,102 @@
+        raise ExpiredTokenError
+
+    # User of device has already made their decision for this device
+    if device.status in (device.DENIED, device.AUTHORIZED):
```
♻️ similar to [this](https://github.com/jazzband/django-oauth-toolkit/pull/1539/files#r2021309632) other comment, worth considering the condition `if device.status != device.AUTHORIZATION_PENDING`

cristiprg replied:
> Actually if this is the user facing confirmation view, I'm feel like these should be errors displayed on the form. 

Just did this fix ✨ 

> user with a valid user code at this juncture.

I would think that a user code is valid only device is `AUTHORIZATION_PENDING`. Otherwise you shouldn't be able to do anything with the use code IMO 🤔

cristiprg commented on oauth2_provider/views/base.py:
```diff
@@ -307,6 +317,45 @@ def post(self, request, *args, **kwargs):
+    def device_flow_token_response(
+        self, request: http.HttpRequest, device_code: str, *args, **kwargs
+    ) -> http.HttpResponse:
+        device = Device.objects.get(device_code=device_code)
```
I checked the RFC [here](https://datatracker.ietf.org/doc/html/rfc8628#section-3.5) and it doesn't really specify how to respond but I think good candidates are `access_denied` or `invalid_request` as defined [here](https://datatracker.ietf.org/doc/html/rfc6749#section-5.2) `or is otherwise malformed.`

cristiprg commented on oauth2_provider/views/device.py:
```diff
@@ -0,0 +1,102 @@
+# device login we're making the decision here to require being logged in
+# up front
+@login_required
+def device_user_code_view(request):
```
more of a question rather than suggestion - as described [here](https://github.com/jazzband/django-oauth-toolkit/pull/1539/files#diff-81e62c4239be18d2d3320260122965d6a179a9518f08ec95fab24e5233f4a955R63) and [here](https://datatracker.ietf.org/doc/html/rfc8628#section-5.1), do we provide any suggestions/instructions/information/reminders to anyone who uses this library that they `SHOULD` implement rate-limiting themselves? 

I believe it is out of the scope for this library to provide rate-limiting, as that's an entire rabbit whole on its own 😂 but still good to mention it somewhere (even in the docs perhaps? 🤔 )

cristiprg commented on oauth2_provider/views/device.py:
```diff
@@ -0,0 +1,102 @@
+        return render(request, "oauth2_provider/device/user_code.html", {"form": form})
+
+    user_code: str = form.cleaned_data["user_code"]
+    device: Device = get_device_model().objects.get(user_code=user_code)
```
[sorry, a 3rd thread of comments to the same line] 

The RFC recommends [here](https://datatracker.ietf.org/doc/html/rfc8628#section-6.1) here: `One way to improve input speed is to restrict the character set to case-insensitive A-Z characters, with no digits.`. I notice here you already partially applied this `character_space = "0123456789ABCDEFGHIJKLMNOPQRSTUV"`, so I'm wondering whether we should transform user input to uppercase before performing the db query?

danlamanna replied:
see also point 3 here: https://github.com/jazzband/django-oauth-toolkit/pull/1539#issuecomment-2582181625.

cristiprg replied:
Thanks @danlamanna! 

I think the code generation is the first half of the story, but the comparison of two tokens is the second half. For example, if we'd like the the code to be case-insensitive, we'd have to modify user input, which I would guess has to be "paired" with the generator (as in, the two have to make sense together).

> I'm being core RFC focused here first and if anything opinionated needs to be added I think we can add it later, This pr is already chunky as is the way I see it. Nothing stopping us from releasing inceremental updates here instead of one big bang :)

However, I agree with @duzumaki  that this could left for a subsequent PR, which I totally support 👍

danlamanna commented:
Any update on this?

pennersr commented on oauth2_provider/views/device.py:
```diff
@@ -0,0 +1,102 @@
+
+
+@login_required
+def device_confirm_view(request: http.HttpRequest, device_code: str):
```
The device code (`device_code`) should be used by the device to poll for the token. The user code (`user_code`) should be confirmed by the user, but I don't see any handling of `user_code` here?

dopry commented on docs/tutorial/tutorial_06.rst:
```diff
@@ -0,0 +1,100 @@
+--------------
+Note: You should already have the `/token` endpoint implemented in your authorization server before this.
+
+Send the following request (in the real world, the device makes this request):
```
Would this be to demonstrate the alternative response?  Would the suggestion here be to add something like if  you send this request and the user hasn't yet accepted then you get response 

...

if the user has accepted you will get the response

...

?

cristiprg replied:
yup, exactly. Just for completeness

dopry commented on pyproject.toml:
```diff
@@ -36,7 +36,7 @@ classifiers = [
 	"django >= 4.2",
 	"requests >= 2.13.0",
-	"oauthlib >= 3.2.2",
+	"oauthlib @ git+https://github.com/oauthlib/oauthlib.git@master",
```
Do we have an official oauthlib release with this functionality yet?

cristiprg replied:
Not yet unfortunately, but I asked here https://github.com/oauthlib/oauthlib/issues/897.

JonathanHuot replied:
Release 3.3.0 is out

cristiprg commented:
Thanks @dopry for the review! Hope you're well ✨  I will pick up the work for this PR tomorrow 💪

dopry commented:
@cristiprg  huzzah, just wanted to let you know this is still on my radar and is the highest priority issue for me to merge. At the moment i'm blocked by lack of code coverage verification with codecov, which is in turn blocked by the projects membership in jazzband (we can update our secrets to set a repo token to get around the organization wide rate limiting). @jezdez said he would help do the transfer on two occasions. I've been waiting for a month now.  Hopefully they'll actually get around to completing the transfer so I can proceed with fixing up our CI and starting to merge these outstanding PRs.

cristiprg commented:
Thanks @dopry for the keeping me in the loop! Hopefully it can be resolved soon 🤞 

Just as a note on code coverage, I saw it around two weeks ago, we were just under the accepted threshold, there was one function without a test. so I added this commit https://github.com/jazzband/django-oauth-toolkit/pull/1539/commits/8470bc31df3d08008c8e2dcbfedc730652f9015b.

I agree it's super useful info to have 🤞

adamcharnock commented on oauth2_provider/views/device.py:
```diff
@@ -0,0 +1,197 @@
+    def post(self, request, *args, **kwargs):
+        headers, response, status = self.create_device_authorization_response(request)
+
+        device_request = DeviceRequest(client_id=request.POST["client_id"], scope=request.POST.get("scope"))
```
Minor request (as someone trying to deploy and test this at the moment on an internal project):

As it stands now, the server will issue a 500 error if it receives a request with a missing `client_id`.

A 400 error would be ideal, eg:

```python
if not request.POST.get("client_id"):
    return http.JsonResponse({'error': 'missing_client_id'}, status_code=400)
```

(or similar). 

Thank you!

cristiprg replied:
Thanks @adamcharnock! I also noticed that but the real fix should be in oauthlib itself

cristiprg replied:
Ooh I see what's going on. I thought for a second there's a bug in oauthlib [here](https://github.com/oauthlib/oauthlib/blob/0667a391a0239a66b977c2d65d5fb52c4d825a34/oauthlib/oauth2/rfc8628/endpoints/device_authorization.py#L129), but here we should either check the status before using `POST["client_id"]`, or simply do the check like you suggested

cristiprg commented:
> @cristiprg we've got the repo moved... Could you rebase and address the `KeyError` issue raised by @adamcharnock? Check and see if the spec has anything particular to say about what to do when client_id isn't provided. 

Sure thing, I'll look into it Wednesday ✨

adamcharnock commented:
To save you some legwork @cristiprg (mostly written by Opus, but I've manually checked the specs and it is indeed correct): 

According to RFC 8628, if the client doesn't provide a client_id when it's required, the server should respond with an error.

From Section 3.1, the client_id is "REQUIRED if the client is not authenticating with the authorization server as described in
Section 3.2.1. of [RFC6749]."

Section 3.2 states: "In the event of an error (such as an invalidly configured client), the authorization server responds in the
same way as the token endpoint specified in Section 5.2 of [RFC6749]."

Per RFC 6749 Section 5.2, this would be a 400 Bad Request with an "invalid_request" error:

```
HTTP/1.1 400 Bad Request
Content-Type: application/json
Cache-Control: no-store

{
  "error": "invalid_request",
  # This value of this "error_description" field is not in the spec
  "error_description": "Missing required parameter: client_id" 
}
```

dopry commented:
Looks like we still have about 5 branches to cover to get full coverage for the tests.

magicbrothers commented:
Great work, I could successfully test it with our downstream application and it „just worked”!

When I entered the user code I noticed that the input has to match the code exactly, as in it may not contain a `-` for readability and it is case-sensitive although the character space only includes (numbers and) uppercase letters.

My suggestion is to follow the recommendations described in [Section 6.1](https://www.rfc-editor.org/rfc/rfc8628#section-6.1) and reduce the character set to `BCDFGHJKLMNPQRSTVWXZ` to avoid confusion with 0 and O, and to strip all characters not in the set before validation, as well as upper-casing the input, to conviniently match inputs like `abcd-efgh` to the code `ABCDEFGH`.

magicbrothers commented:
@dopry what's the status with this PR? Is there a chance for it to be merged and released soon?

dopry commented:
It would be awesome if you could setup the tests/app/idp to have a view for testing the flow...  It would save me needing to set something up for the e2e tests.

dopry commented:
I might have meant the RP... I'll try to test next week.

cristiprg commented:
@dopry 
I tried the new demo page on my local and it looks really nice, and it explains well all the steps for the device authorization flow 🙌 Frontend JS/Svelte is not my strong point tbh, so I cannot give too constructive feedback but at a high level the code itself seems good to me 👍 

My only concern is that `if the confirmation device is going to be a web browser`, then in the real world no one should even consider using the device authorization flow as it's not meant for that use case (instead it'll probably be `authorization-code`). I think you have explained the use case well on the main page, that the device authorization flow `is used by devices with limited input capabilities (like smart TVs, IoT devices, etc.) to obtain OAuth tokens.`, however I think we perhaps need an extra comment, such as "do not use device-authorization if the device is a browser, this is just an illustrative example how you'd need to implement the flow on your device", just to be extra sure no one gets confused. 

And a similar comment next to the CORS configuration in the JS code, that you should never really need to add those URLs to your CORS allow origin configuration. This is important because people copy/pasta a lot, which is normal, but you wouldn't want to copy/pasta that 😁 

Excitiiiiing!

dopry commented:
Good catch..  I was kinda blinded by the work I was doing to get the demo working to think about the broader CORS context. 

I'll try to address that.

No worries on the svelte stuff. I'm quick enough at it

dopry commented:
@duzumaki @cristiprg @danlamanna @adamcharnock @magicbrothers, et al.. I wanted to say thank you to everyone who participated in the implemntation and review of this feature. I'll work on a release for it this week.
